File: java/ecbenchmark/src/ecbenchmark/cec10/F14.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package ecbenchmark.cec10;

import java.util.ArrayList;
import java.util.List;
import ecbenchmark.function.Elliptic;
import ecbenchmark.Function;
import ecbenchmark.operator.Grouped;
import ecbenchmark.operator.Permuted;
import ecbenchmark.operator.Shifted;

/**
 *
 * @author jcrada
 */
public class F14 extends CecFunction {

    private List<Function> sumElliptic = new ArrayList<Function>();

    public F14(int dimensions, int mValue) {
        super("F14", "(D/m)-group Shifted and m-rotated Elliptic's Function", dimensions,
                -100, 100, true, new CecRandom(14l), mValue);

        double[] shift = CecMath.ShiftVector(dimensions, getMinimum(), getMaximum(), getRandomizer());
        int[] permutation = CecMath.PermutationVector(dimensions, getRandomizer());
        double[][] rotation = CecMath.RotationMatrix(mValue, getRandomizer());
        double[] permutedShift = new double[dimensions];
        for (int i = 0; i < dimensions; ++i) {
            permutedShift[i] = shift[permutation[i]];
        }
        for (int k = 0; k < dimensions / mValue; ++k) {
            sumElliptic.add(
                    new Permuted(permutation, new Grouped(k * mValue, (k + 1) * mValue,
                    new Shifted(permutedShift, k * mValue, new CecRotated(rotation, new Elliptic())))));
        }

    }

    public double f(final double[] x) {
        double result = 0;
        for (Function rotatedElliptic : sumElliptic) {
            result += rotatedElliptic.f(x);
        }
        return result;
    }
}

File: cpp/ecbenchmark/src/cec10/F14.cpp
#include "ecbenchmark/cec10/F14.h"
#include "ecbenchmark/cec10/CecMath.h"
#include "ecbenchmark/operator/Permuted.h"
#include "ecbenchmark/operator/Grouped.h"
#include "ecbenchmark/operator/Shifted.h"
#include "ecbenchmark/cec10/CecRotated.h"
#include "ecbenchmark/function/Elliptic.h"

namespace ecb {
    namespace cec10 {

        F14::F14(int dimensions, int mValue)
        : CecFunction("F14", "(D/m)-group Shifted and m-rotated Elliptic's Function", dimensions,
        -100, 100, true, mValue, new CecRandom(14l)) {

            std::vector<scalar>  shift = CecMath::ShiftVector(dimensions, minimumDomain(), maximumDomain(), randomizer());
            std::vector<int>  permutation = CecMath::PermutationVector(dimensions, randomizer());
            std::vector<std::vector<scalar> >  rotation = CecMath::RotationMatrix(mValue, randomizer());
            std::vector<scalar>  permutedShift (dimensions,0);
            for (int i = 0; i < dimensions; ++i) {
                permutedShift[i] = shift[permutation[i]];
            }
            for (int k = 0; k < dimensions / mValue; ++k) {
                sumElliptic.push_back(
                        new Permuted(permutation, new Grouped(k * mValue, (k + 1) * mValue,
                        new Shifted(permutedShift, k * mValue, new CecRotated(rotation, new Elliptic())))));
            }

        }

        F14::~F14() {
            for (size_t i = 0 ; i < sumElliptic.size(); ++i){
                delete sumElliptic[i];
            }
        }

        scalar F14::f(const std::vector<scalar>& x) {
            scalar result = 0;
            for (size_t i = 0 ; i < sumElliptic.size() ; ++i) {
                result += sumElliptic[i]->f(x);
            }
            return result;
        }
    }
}

